### PR TITLE
[Payments pallet] - Create, release and cancel / Fee Handler

### PR DESCRIPTION
**Pull request by darkforest0202**
Related #280 #284 #247 

Phase 1: Basic functionality
Use case: A buys a product from B using the pallet and a payment is placed using the pay functionality. At this point, balance is transferred to B but frozen. This allows B to see the new balance but not available for it's use until A confirms the product was received.
Then when A confirms that product was received, it free the balance by calling release. cancel is trivial.

pay -> (add payment_detail type (enum) )
release
cancel

---

Also an enhanced version of the fees was introduced:

```rust
#[derive(Clone, Encode, Decode, Eq, PartialEq, Default, MaxEncodedLen, TypeInfo, Debug)]
pub struct Fees<BoundedFeeDetails> {
	pub sender_pays: BoundedFeeDetails,
	pub beneficiary_pays: BoundedFeeDetails,
}
```

where `BoundedFeeDetails` could be typed like this:
```rust
pub type BoundedFeeDetails<T> =
	BoundedVec<(<T as frame_system::Config>::AccountId, BalanceOf<T>), <T as Config>::MaxFees>;
```

and the `FeeHandler` now retrieves:

```
pub trait FeeHandler<T: pallet::Config, BoundedFeeDetails> {
	/// Get the distribution of fees to marketplace participants
	fn apply_fees(
		sender: &T::AccountId,
		beneficiary: &T::AccountId,
		amount: &BalanceOf<T>,
		remark: Option<&[u8]>,
	) -> Fees<BoundedFeeDetails>;
}
```

Next steps:

- Add Ids to the Payments storage item being a `StorageNMap` type with keys : sender, beneficiary, payment_id
- Add benchmarking.

### CODE REVIEW

**Comment by darkforest0202** [comment]
I need to solve some clippy complains btw.

File: pallets/payments/src/types.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+/// Fee Handler trait that defines how to handle marketplace fees to every
+/// payment/swap
+pub trait FeeHandler<T: pallet::Config, BoundedFeeDetails> {
+	/// Get the distribution of fees to marketplace participants
+	fn apply_fees(
+		sender: &T::AccountId,
+		beneficiary: &T::AccountId,
+		amount: &BalanceOf<T>,
```
Comment (olanod) [inline pallets/payments/src/types.rs:63]: AssetId is probably necessary for fee handlers to determine the right fees

**Review by olanod — CHANGES_REQUESTED** [review]
Great job! I'm excited to see the pallet in action.
Left several comments for changes that were easier to realize once seeing the code.

File: pallets/payments/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+	/// The function will create a new payment. The fee and incentive
+	/// amounts will be calculated and the `PaymentDetail` will be added to
+	/// storage.
+	fn create_payment(
+		sender: &T::AccountId,
+		beneficiary: &T::AccountId,
+		asset: AssetIdOf<T>,
+		amount: BalanceOf<T>,
+		payment_state: PaymentState<BlockNumberFor<T>>,
+		incentive_percentage: Percent,
```
[the remark this answers is not in the document]
Comment (darkforest0202) [reply pallets/payments/src/lib.rs:296]: hmm what would be the gain of making this change?

**Review by olanod — APPROVED** [review]

**Pull request merged**